azdev/operations/setup.py

```diff
@@ -258,14 +264,114 @@ def add_ext_repo(path):
+                           " and a cli repo must be specified")
+        if any([use_global, copy, set_env]):
+            raise CLIError("if global, copy, or set env are set then both an"
+                           " extensions repo and a cli repo must be specified")
+
 
-    require_virtual_env()
+def _check_paths(cli_path, ext_repo_path):
+    if not os.path.isdir(cli_path):
+        raise CLIError("The cli path is not a valid directory, please check the path")
+    if not os.path.isdir(ext_repo_path):
+        raise CLIError("The cli extensions path is not a valid directory, please check the path")
 
-    start = time.time()
+
+def setup(cli_path=None, ext_repo_path=None, ext=None, deps=None, set_env=None, copy=None, use_global=None):
+    if not set_env:
+        if not get_env_path():
+            raise CLIError('You are not running in a virtual enviroment and have not chosen to set one up.')
+    elif 'VIRTUAL_ENV' in os.environ:
+        raise CLIError("You are already running in a virtual enviroment, yet you want to set a new one up")
 
     heading('Azure CLI Dev Setup')
 
+    _validate_input(cli_path, ext_repo_path, set_env, copy, use_global)
+    # cases for handling legacy install
+    if not any([cli_path, ext_repo_path]) or cli_path == "pypi" or (not cli_path or not ext_repo_path):
+        return _handle_legacy(cli_path, ext_repo_path, ext, deps, time.time())
+    if 'CONDA_PREFIX' in os.environ:
+        raise CLIError('CONDA virutal enviroments are not supported outside'
+                       ' of interactive mode or when -c and -r are provided')
+    _check_paths(cli_path, ext_repo_path)
+
+    if set_env:
+        subprocess.call(shlex.split((const.VENV_CMD if const.IS_WINDOWS else const.VENV_CMD3) +
+                                    set_env), shell=False)
+        azure_path = os.path.join(os.path.abspath(os.getcwd()), set_env)
+    else:
+        azure_path = os.environ.get(const.VIRTUAL_ENV)
+
+    dot_azure_config = os.path.join(azure_path, '.azure')
+    dot_azdev_config = os.path.join(azure_path, '.azdev')
+
+    # clean up venv dirs if they already existed
+    # and this is a reinstall/new setup
+    if os.path.isdir(dot_azure_config):
+        shutil.rmtree(dot_azure_config)
+    if os.path.isdir(dot_azdev_config):
+        shutil.rmtree(dot_azdev_config)
+
+    global_az_config = os.path.expanduser(os.path.join('~', '.azure'))
+    global_azdev_config = os.path.expanduser(os.path.join('~', '.azdev'))
+    azure_config_path = os.path.join(dot_azure_config, const.CONFIG_NAME)
+    azdev_config_path = os.path.join(dot_azdev_config, const.CONFIG_NAME)
+
+    if os.path.isdir(global_az_config) and copy:
+        shutil.copytree(global_az_config, dot_azure_config)
+        if os.path.isdir(global_azdev_config):
+            shutil.copytree(global_azdev_config, dot_azdev_config)
+        else:
+            os.mkdir(dot_azdev_config)
+            file = open(azdev_config_path, "w")
```

Comment:
why need open it in write mode?

Reply:
This is just for creating the azdev config file if it does not exist.

Reply:
would you use `with` statement?